### PR TITLE
Generator seed

### PR DESCRIPTION
**Pull request by mh-nguyen712**
### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

Hi deepinv team. I add a new feature to `PhysicsGenerator` class. Each class will have now a `rng` attribute for random number generator for reproducibility, as suggested in #243 . Moreover, we can also specify the `seed` in each generation step. I also fixed the `ProductConvolutionBlurGenerator` for generating a batch of eigen-PSF, which is not supported previously, and enhance the SVD quality. 
An another small feature is that I created an equivalent of `numpy.random.choice` in torch, which allows `PhysicsGenerator` from inpainting and mri to take torch.Generator instead of numpy.Generator, make it more uniform and avoid some overhead of switching between torch and numpy, mayby @Andrewwango  can take a look?

Don't hesitate to make any comments!

### CODE REVIEW

**Comment by Andrewwango** [comment]
Hi @minh-hai-nguyen-712, thank you for the PR! I will review it when I return from annual leave w/c Mon 26th.

File: deepinv/optim/fixed_point.py [hunk 1]
```diff
@@ -242,7 +242,6 @@ def forward(self, *args, compute_metrics=False, x_gt=None, **kwargs):
             range(self.max_iter),
             disable=(not self.verbose or not self.show_progress_bar),
         ):
-
```
Comment (Andrewwango) [inline deepinv/optim/fixed_point.py:245]: Why are there lots of changed files with newlines removed?

Comment (minh-hai-nguyen-712) [reply deepinv/optim/fixed_point.py:245]: It's was due to black formatter. Also, as I made some improvements on SpaceVaryingBlurGenerator, which lead to some neccesary changes in SpaceVaryingBlur physics. The changes is mainly  about the documentation.

Comment (minh-hai-nguyen-712) [reply deepinv/optim/fixed_point.py:245]: And I forgot to turn off my autopep8 formatter so sometimes it changes `lambda x: ....` into `def func(x): ...`

File: deepinv/physics/generator/base.py [hunk 2]
```diff
@@ -26,41 +29,75 @@ class PhysicsGenerator(nn.Module):
...
+        self.device = device
+        if rng is None:
+            self.rng = torch.Generator(device=device)
+        else:
+            # Make sure that the random generator is on the same device as the physic generator
+            assert (
+                rng.device == device
+            ), "The random generator is not on the same device as the Physic Generator"
+            self.rng = rng
+        self.initial_random_state = self.rng.get_state()
```
Comment (Andrewwango) [inline deepinv/physics/generator/base.py:65]: Can we have a `reset_rng` method that sets the rng back to the initial state? That way, we can have reproducibility in epochs for #202 (provided that `shuffle=False` in `DataLoader`). And add a test to test this.

Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/base.py:65]: Good idea. I'll try to do this as soon as possible.

Comment (Andrewwango) [reply deepinv/physics/generator/base.py:65]:
```python
from torch.utils.data import DataLoader
import torch
from deepinv.tests.dummy_datasets.datasets import DummyCircles
from deepinv.physics import Inpainting
from deepinv.physics.generator import BernoulliSplittingMaskGenerator
from deepinv.utils import plot

# Sample dataset
img_size = (3, 8, 8)
dataset = DummyCircles(5, imsize=img_size)
loader = DataLoader(dataset, shuffle=False)

# Random inpainting physics
physics = Inpainting(tensor_size=img_size) # No mask for now
physics_generator = BernoulliSplittingMaskGenerator(tensor_size=img_size, split_ratio=0.5)

# Simulate a training loop for 2 epochs
out = []
for epoch in range(2):
    physics_generator.reset_rng()
    
    ys = []
    mask_prev = torch.tensor([0])
    for x in loader:
        params = physics_generator.step()
        y = physics(x, **params)
        ys += [y]
        # Test masks different every iteration
        assert not torch.equal(params["mask"], mask_prev)
        mask_prev = params["mask"]

    out += [torch.cat(ys, dim=-2)]

out = torch.cat(out, dim=-1)

# Test ys same every epoch
assert torch.equal(out[..., :8], out[..., 8:])

# Optional visualisation
plot(out)
```

File: deepinv/optim/fixed_point.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/optim/fixed_point.py:245]: Please can we remove the extra newlines changes as they aren't relevant to this PR?

File: deepinv/tests/test_physics.py [hunk 3]
```diff
@@ -580,13 +590,14 @@ def test_noise(device, noise_type):
...
     """
-    physics = dinv.physics.DecomposablePhysics()
+    physics = dinv.physics.DecomposablePhysics(device=device)
     physics.noise_model = choose_noise(noise_type)
     x = torch.ones((1, 3, 2), device=device).unsqueeze(0)
 
     y1 = physics(
         x
-    )  # Note: this works but not physics.A(x) because only the noise is reset (A does not encapsulate noise)
+        # Note: this works but not physics.A(x) because only the noise is reset (A does not encapsulate noise)
```
Comment (Andrewwango) [inline deepinv/tests/test_physics.py:599]: And revert these stylistic changes caused by the formatter (and the lambda thing above)

File: deepinv/physics/generator/inpainting.py [hunk 4]
```diff
@@ -301,14 +308,15 @@ def batch_step(self, input_mask: torch.Tensor = None) -> dict:
 
         for c in range(_C):
             for t in range(_T):
-                ind = self.rng.choice(
+                ind = random_choice(
```
Comment (Andrewwango) [inline deepinv/physics/generator/inpainting.py:311]: Thank you so much for making these changes!!

File: deepinv/optim/fixed_point.py (hunk 1 above)
Comment (minh-hai-nguyen-712) [reply deepinv/optim/fixed_point.py:245]: The extra changes came from when I merged this branch with main. I'll remove it as you suggested

File: deepinv/physics/functional/random.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+from typing import Tuple
+from warnings import warn
+
+import numpy as np
+import torch
+
+
+def random_choice(
```
Comment (Andrewwango) [inline deepinv/physics/functional/random.py:8]: I know it's a bit laborious, but please could you add a test for this function? Maybe just a hardcoded test to test the rng (i.e. `random_choice(...) == torch.Tensor([...])`, and something to test the replacement and p, like:
```python
a = torch.Tensor([1,2,3,4,5])
p = torch.Tensor([0,0,1,0,0])
random_choice(a, 2, replace=True, p=p) 
```
should equal to [3,3].

Comment (Andrewwango) [reply deepinv/physics/functional/random.py:8]: But thanks for providing this implementation! Is it taken from anywhere?

**Comment by Andrewwango** [comment]
@matthieutrs  please can you review the product blur convolution bits in this? I don't know anything about that!

File: deepinv/physics/functional/random.py (hunk 5 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/functional/random.py:8]: I'll try for the tests. I implemented from scratch with the same logic and error as numpy, there is some examples provided in numpy docs for the case with equal probability. I'm not sure if there exists a better implementation out there.

File: deepinv/physics/generator/mri.py [hunk 6]
```diff
@@ -19,22 +20,22 @@ class BaseMaskGenerator(PhysicsGenerator):
     :param Tuple img_size: image size, either (H, W) or (C, H, W) or (C, T, H, W), where optional C is channels, and optional T is number of time-steps
     :param int acceleration: acceleration factor, defaults to 4
     :param float center_fraction: fraction of lines to sample in low frequencies (center of k-space). If 0, there is no fixed low-freq sampling. Defaults to None.
-    :param np.random.Generator rng: numpy random generator, defaults to np.random.default_rng()
+    :param torch.Generator rng: torch random generator, defaults to None
```
Comment (matthieutrs) [inline deepinv/physics/generator/mri.py:23]:
```suggestion
    :param torch.Generator rng: torch random generator. Defaults to None.
```

File: deepinv/tests/test_generators.py [hunk 7]
```diff
@@ -303,14 +323,778 @@ def test_generation(name, device):
                 ]
             ).to(device)
 
+    elif name == "ProductConvolutionBlurGenerator":
+        w = params["filters"]
+        # atol = 1e-5
+        atol = 1e-2
```
Comment (matthieutrs) [inline deepinv/tests/test_generators.py:329]: Isn't this a bit too high?

File: deepinv/tests/test_generators.py [hunk 8]
```diff
@@ -93,20 +112,21 @@ def test_generation_newparams(name, device):
...
+        param_key = ["filter"]
+    elif name == "ProductConvolutionBlurGenerator":
+        param_key = ["filters", "multipliers"]
     elif name == "SigmaGenerator":
-        param_key = "sigma"
+        param_key = ["sigma"]
 
-    params0 = generator.step(batch_size=batch_size)
-    params1 = generator.step(batch_size=batch_size)
+    params0 = generator.step(batch_size=batch_size, seed=0)
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply deepinv/tests/test_generators.py:125]: I did not find it either

**Review by matthieutrs — COMMENTED** [review]
Thanks a lot for this amazing PR!

File: deepinv/physics/functional/interp.py [hunk 9]
```diff
@@ -179,15 +203,14 @@ def _radial_distance(self, X: torch.Tensor) -> torch.Tensor:
         return dist**2 * torch.log(dist)
 
 
-def _ensure_2d(tensor: torch.Tensor) -> torch.Tensor:
+def _ensure_4d(tensor: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/physics/functional/interp.py:206]: Yes your suggestion is better. I'll change it

File: deepinv/physics/generator/base.py [hunk 10]
```diff
@@ -26,41 +29,81 @@ class PhysicsGenerator(nn.Module):
...
+
+        :param int seed: the seed to set for the random number generator.
+         If not provided, the current state of the random number generator is used.
+         Note: The `torch.manual_seed` is triggered when a the random number generator is not initialized.
+        """
+        if seed is not None:
+            if self.rng is not None:
+                self.rng = self.rng.manual_seed(seed)
+            else:
+                torch.manual_seed(seed)
```
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/base.py:97]: Currently all randomness in our library come from torch. torch.manual_seed() will set the random seed for all devices. Your function is good for seed everything but It might cause some problems for other randomness? Also, in the __init__, if a rng is not provided, a new rng on the same device will be created under the hood so the call to torch.manual_seed() might disappear.

File: deepinv/tests/test_generators.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (minh-hai-nguyen-712) [reply deepinv/tests/test_generators.py:125]: I'll add this test

File: deepinv/tests/test_generators.py (hunk 7 above)
Comment (minh-hai-nguyen-712) [reply deepinv/tests/test_generators.py:329]: It passed on my 2 local machines with atol=1e-5 with newest torch version. But it doesn't pass on github.
I'm trying to figure out the problem. It would be great if you can test it on your local machine.

**Comment by minh-hai-nguyen-712** [comment]
> Thanks a lot for this amazing PR!

Thank you very much for your detailed review!

File: deepinv/physics/generator/base.py (hunk 2 above)
Comment (minh-hai-nguyen-712) [reply deepinv/physics/generator/base.py:65]: The issue #202 is solved now. To have the expected behavior, it's enough to call physiscs_generator.reset_rng() at the beginning of each epoch.

File: deepinv/tests/test_generators.py (hunk 7 above)
Comment (minh-hai-nguyen-712) [reply deepinv/tests/test_generators.py]: It's weird for me. I tested it on multiple local machine and it passed every time. But it never passed on github.

File: deepinv/physics/generator/base.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/physics/generator/base.py:65]: Ok, I added the appropriate warning to close #202

File: deepinv/tests/test_generators.py (hunk 7 above)
Comment (mh-nguyen712) [reply deepinv/tests/test_generators.py]: @matthieutrs Do you know on which hardware Github runs our tests? I ran the test on all of my local machine, I also tried on Google Colab and it passed every time.

Comment (mh-nguyen712) [reply deepinv/tests/test_generators.py]: All test passed now. Some of PhysicsGenerator requires complex operation like torch.svd, which give different values depending on hardware, torch version, etc, so for the test, I only test on float64 with atol=1e-8.

**Comment by Andrewwango** [comment]
Thanks for your work @mh-nguyen712 ! I have just added another issue #297  to also set rngs in physics which have randomness - if you are interested in combining with this PR!

**Comment by mh-nguyen712** [comment]
> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!



> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!

Of course! I can combine it to this PR!

**Comment by mh-nguyen712** [comment]
> Thanks for your work @mh-nguyen712 ! I have just added another issue #297 to also set rngs in physics which have randomness - if you are interested in combining with this PR!

I have added rng to all noise model, except for `GammaNoise`, which depends on torch.distributions and dont have support of `torch.Generator`. All tests passed. I also added some small code improvement (e.g., use new `random_choice` function). Could you review them?

File: deepinv/physics/generator/base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/physics/generator/base.py]: I agree with @mh-nguyen712 - I think our code should remain functional and not interfere with the global state - if the user really wants to set a universal seed maybe we can add the function that Matthieu suggested in `deepinv.utils` as a utility function?

**Comment by Andrewwango** [comment]
I'm happy to approve - @matthieutrs ?

**Review by matthieutrs — COMMENTED** [review]
Thanks for this great PR! Just reoppeing some minor comments which were closed but not taken into account it seems

File: deepinv/tests/test_generators.py (hunk 7 above)
Comment (matthieutrs) [reply deepinv/tests/test_generators.py]: Thanks!

File: deepinv/physics/generator/base.py [hunk 11]
```diff
@@ -26,41 +29,81 @@ class PhysicsGenerator(nn.Module):
...
         super().__init__()
 
         self.step_func = step
         self.kwargs = kwargs
         self.factory_kwargs = {"device": device, "dtype": dtype}
+        self.device = device
+        if rng is None:
+            self.rng = torch.Generator(device=device)
+        else:
+            # Make sure that the random generator is on the same device as the physic generator
```
[the remark this answers is not in the document]
Comment (matthieutrs) [reply deepinv/physics/generator/base.py:60]: @mh-nguyen712 I think this would actually be better with `physics`

[the remark this answers is not in the document]
Comment (mh-nguyen712) [reply deepinv/physics/generator/base.py:60]: Sorry I only changed physic to physics in the warning message but forgot the comment.

File: deepinv/physics/noise.py [hunk 12]
```diff
@@ -177,11 +277,13 @@ class GammaNoise(torch.nn.Module):
     Distribution for modelling speckle noise (eg. SAR images),
     where :math:`\ell>0` controls the noise level (smaller values correspond to higher noise).
 
+    This noise model does not support the random number generator.
```
[the remark this answers is not in the document]
Comment (mh-nguyen712) [reply deepinv/physics/noise.py]: Added

**Comment by mh-nguyen712** [comment]
Sorry about that. I just resolved the open comments!

**Review by matthieutrs — APPROVED** [review]
